relayer/src/kill_switch.rs

```diff
@@ -52,161 +40,180 @@ impl_metered_service! {
+    ChallengeRoot { tx_hash: Option<TxHash> },
+    // Waiting for submit merkle root tx to be finalized
+    SubmitMerkleRoot { tx_hash: Option<TxHash> },
+    // Exit
+    Exit,
 }
 
 pub struct KillSwitchRelayer {
     api_provider: ApiProviderConnection,
     eth_api: EthApi,
-    genesis_config: GenesisConfig,
-    proof_storage: Arc<dyn ProofStorage>,
+    http_client: HttpClient,
+    relayer_http_url: String,
 
     start_from_eth_block: Option<u64>,
     state: State,
-    block_finality_storage: sled::Db,
+    challenged_block: Option<u64>,
 
     metrics: Metrics,
-    block_finality_metrics: BlockFinalityArchiverMetrics,
-
-    count_thread: Option<usize>,
 }
 
 impl MeteredService for KillSwitchRelayer {
     fn get_sources(&self) -> impl IntoIterator<Item = Box<dyn prometheus::core::Collector>> {
-        self.metrics
-            .get_sources()
-            .into_iter()
-            .chain(self.block_finality_metrics.get_sources())
-            .chain(prover_interface::Metrics.get_sources())
+        self.metrics.get_sources()
     }
 }
 
 impl KillSwitchRelayer {
     pub async fn new(
         api_provider: ApiProviderConnection,
         eth_api: EthApi,
-        genesis_config: GenesisConfig,
-        proof_storage: Arc<dyn ProofStorage>,
+        http_client: HttpClient,
         from_eth_block: Option<u64>,
-        block_finality_storage: sled::Db,
-
-        count_thread: Option<usize>,
+        relayer_http_url: String,
     ) -> Self {
         Self {
             api_provider,
             eth_api,
-            genesis_config,
-            proof_storage,
+            relayer_http_url,
+            http_client,
             start_from_eth_block: from_eth_block,
-            state: State::Normal,
-            block_finality_storage,
+            state: State::ScanForEvents,
+            challenged_block: None,
             metrics: Metrics::new(),
-            block_finality_metrics: BlockFinalityArchiverMetrics::new(),
-            count_thread,
         }
     }
 
     pub async fn run(&mut self) -> anyhow::Result<()> {
-        self.spawn_block_finality_archiver()?;
-
         log::info!("Starting kill switch relayer");
 
-        let mut attempts = 0;
-
         loop {
-            attempts += 1;
-            let now = Instant::now();
+            let loop_delay;
+
             let res = match &self.state {
-                State::Normal => self.main_loop().await,
-                State::WaitingForKillSwitchTxFin { tx_hash, .. } => {
-                    self.check_kill_switch_tx_finalized(*tx_hash).await
+                State::ScanForEvents => self.scan_for_events().await.map_err(Error::from),
+                State::ChallengeRoot { .. } => self.challenge_root().await.map_err(Error::from),
+                State::SubmitMerkleRoot { .. } => {
+                    self.submit_merkle_root().await.map_err(Error::from)
+                }
+                State::Exit => {
+                    process::exit(1);
```

Comment:
cannot we simply return?

Reply:
Yes, if we don't need erroneous return code